### PR TITLE
feat: article collect on mirror

### PR DESCRIPTION
**Pull request by nuanyang233**
# Pull Request

## Related Issue

Closes FW-1334

## Checklist

-   [ ] I have tested these changes locally.
-   [ ] I have reviewed the code changes.
-   [ ] I have added unit tests if applicable.

### CODE REVIEW

File: src/app/api/paragraph/route.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    const link = searchParams.get('link');
+    if (!link) return Response.json({ error: 'Missing link' }, { status: 400 });
+
+    const response = await fetch(link, {
+        headers: { 'User-Agent': 'Twitterbot' },
+        signal: request.signal,
+    });
+
+    if (!response.ok || (response.status >= 500 && response.status < 600)) {
+        return Response.json({ error: 'Unable to digest link' }, { status: 500 });
```
Comment (guanbinrui) [inline src/app/api/paragraph/route.ts:18]:
```suggestion
        return Response.json({ error: `Unable to digest link = ${link}` }, { status: response.status });
```

Comment (guanbinrui) [inline src/app/api/paragraph/route.ts:18]: FetchError

File: src/app/api/paragraph/route.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import { createSuccessResponseJSON } from '@/helpers/createSuccessResponseJSON.js';
+import { parseJSON } from '@/helpers/parseJSON.js';
+
+export async function GET(request: Request) {
+    const { searchParams } = new URL(request.url);
+    const link = searchParams.get('link');
+    if (!link) return Response.json({ error: 'Missing link' }, { status: 400 });
+
+    const response = await fetch(link, {
+        headers: { 'User-Agent': 'Twitterbot' },
```
Comment (guanbinrui) [inline src/app/api/paragraph/route.ts:13]: Might not necessary.

File: src/app/api/paragraph/route.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            collectorWallet: target.collectibleWalletAddress,
+            position: {
+                from: 0,
+                to: 0,
+            },
+            symbol: data?.props.pageProps.initialState.blog.blog.url?.replace('@', '').toUpperCase(),
+            referrerAddress: data?.props.pageProps.initialState.blog.blog.user.referrerWalletAddress,
+        };
+
+        return createSuccessResponseJSON(result);
```
Comment (guanbinrui) [inline src/app/api/paragraph/route.ts:88]: Please create a Processor for Paragraph links like Frame Processor and use Redis to cache the processed results.

File: src/modals/CollectArticleModal.tsx [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+        const { data, isLoading: queryDetailLoading } = useQuery({
+            enabled: !!props?.article && open,
+            queryKey: ['article', props?.article.platform, props?.article.id, props?.article.origin],
+            queryFn: async () => {
+                if (!props?.article?.origin) return;
+                const provider = resolveArticleCollectProvider(props.article.platform);
+
+                return provider.getArticleDetail(
+                    props.article.platform === ArticlePlatform.Paragraph ? props.article.origin : props.article.id,
```
Comment (guanbinrui) [inline src/modals/CollectArticleModal.tsx:52]: It could fail to acess the correct field silently if a new article provider is added.

Comment (guanbinrui) [inline src/modals/CollectArticleModal.tsx:52]:
```suggestion
                return provider.getArticleDetail(resolveArticleId(props.article),
```

**resolveArticleId.ts**

```ts
export function getArticleId(article: Article) {
    switch (article.platform) {
        // enumerate all the platforms
    }
}
```

File: src/providers/paragraph/index.tsx [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+                method: 'GET',
+            },
+        );
+
+        const data = response.data;
+
+        if (!data) throw new Error('Failed to fetch article detail');
+
+        let isCollected = false;
+        let soldNum = 0;
```
Comment (guanbinrui) [inline src/providers/paragraph/index.tsx:32]:
```suggestion
        let soldCount = 0;
```

File: src/providers/types/Article.ts [hunk 6]
```diff
@@ -58,3 +80,29 @@ export interface Provider {
...
+     */
+    getArticleDetail: (digest: string) => Promise<ArticleCollectDetail>;
+
+    /**
+     * Retrieves the estimated gas fee for collecting an article.
+     * @param detail article detail
+     * @param account user account
+     * @returns
+     */
+    estimateCollectGas: (detail: ArticleCollectDetail, account: string) => Promise<bigint>;
```
Comment (guanbinrui) [inline src/providers/types/Article.ts:99]: Those two members could be as private.

File: src/providers/types/Article.ts [hunk 7]
```diff
@@ -58,3 +80,29 @@ export interface Provider {
...
+
+    /**
+     * Collect an article
+     * @param detail article detail
+     * @param account user account
+     * @returns
+     */
+    collect: (detail: ArticleCollectDetail, account: string) => Promise<bigint>;
```
Comment (guanbinrui) [inline src/providers/types/Article.ts:107]:
```suggestion
    collect: (article: Article, account: string) => Promise<bigint>;
```

File: src/providers/types/Article.ts [hunk 8]
```diff
@@ -58,3 +80,29 @@ export interface Provider {
      */
     getFollowingArticles: (indicator?: PageIndicator) => Promise<Pageable<Article, PageIndicator>>;
 }
+
+export interface ArticleCollectProvider {
```
Comment (guanbinrui) [inline src/providers/types/Article.ts:84]: Merge into `Provider` if you accept all the following suggestions.

File: src/providers/types/Article.ts [hunk 9]
```diff
@@ -33,9 +33,31 @@ export interface Article {
     id: string;
     coverUrl: string | null;
     hasBookmarked?: boolean;
+    slug?: string;
     followingSources: FollowingSource[];
 }
 
+export interface ArticleCollectDetail {
+    quantity?: number;
+    soldNum: number;
```
Comment (guanbinrui) [inline src/providers/types/Article.ts:42]:
```suggestion
    soldCount: number;
```

File: src/app/api/paragraph/route.ts (hunk 3 above)
Comment (nuanyang233) [reply src/app/api/paragraph/route.ts:88]: No, the data (`soldNum` and `proxyAddress`) is frequently updated. When an article is minted for the first time, its collection contract will be created. If we cache this data, we won't be able to get the latest contract address.

File: src/providers/paragraph/index.tsx (hunk 5 above)
Comment (nuanyang233) [reply src/providers/paragraph/index.tsx:32]: Just keep the fields synchronized with the mirror data.

File: src/providers/types/Article.ts (hunk 7 above)
Comment (nuanyang233) [reply src/providers/types/Article.ts:107]: `Article` is used for data display, `ArticleCollectDetail` is used for collecting data, such as `contractAddress`.

File: src/providers/types/Article.ts (hunk 6 above)
Comment (nuanyang233) [reply src/providers/types/Article.ts:99]: Might not necessary.

File: src/providers/mirror/index.tsx
Comment (guanbinrui) [inline src/providers/mirror/index.tsx:1]: index.ts

File: src/providers/types/Article.ts [hunk 10]
```diff
@@ -57,4 +79,28 @@ export interface Provider {
...
+     */
+    estimateCollectGas: (detail: ArticleCollectDetail, account: string) => Promise<bigint>;
+
+    /**
+     * Collect an article
+     * @param detail article detail
+     * @param account user account
+     * @returns
+     */
+    collect: (detail: ArticleCollectDetail, account: string) => Promise<bigint>;
```
Comment (guanbinrui) [inline src/providers/types/Article.ts:105]:
```suggestion
    estimateCollectGas: (detail: ArticleCollectDetail) => Promise<bigint>;

    /**
     * Collect an article
     * @param detail article detail
     * @param account user account
     * @returns
     */
    collect: (detail: ArticleCollectDetail) => Promise<bigint>;
```

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**